test/unit/TestRecipeCoverage.hs

```diff
@@ -0,0 +1,48 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Ensure recipe coverage for all entities that
+-- grant capabilities (aka "devices").
+module TestRecipeCoverage where
+
+import Control.Lens ((^.))
+import Data.Map qualified as M
+import Data.Set qualified as Set
+import Data.Text qualified as T
+import Swarm.Game.Entity (EntityMap (entitiesByCap), entityName)
+import Swarm.Game.Recipe (recipeOutputs)
+import Swarm.TUI.Model (RuntimeState, stdEntityMap, stdRecipes)
+import Swarm.Util (binTuples, commaList, quote)
+import Test.Tasty
+import Test.Tasty.ExpectedFailure (expectFailBecause)
+import Test.Tasty.HUnit
+
+testDeviceRecipeCoverage :: RuntimeState -> TestTree
+testDeviceRecipeCoverage rs =
+  testGroup
+    "Recipe coverage"
+    [ expectFailBecause "Need to come up with more recipes" checkCoverage
+    ]
+ where
+  checkCoverage :: TestTree
+  checkCoverage =
+    testCase
+      "Ensure all devices have recipes (#1268)"
+      $ assertBool errMessage
+      $ null nonCoveredEntities
+   where
+    errMessage =
+      T.unpack $
+        T.unwords
+          [ "Missing recipes for:"
+          , commaList $ map quote $ Set.toList nonCoveredEntities
+          ]
+
+    -- Only include entities that grant a capability:
+    entityNames = Set.fromList . map (^. entityName) . concat . M.elems . entitiesByCap $ rs ^. stdEntityMap
+
+    recipesByOutputEntity = binTuples $ concatMap (\r -> map (,r) $ getOutputsForRecipe r) $ rs ^. stdRecipes
```

Comment:
Why bother constructing a map here by tupling with the recipe and using `binTuples`, when we're just going to later keep only the keys?  Why not just
```
outputEntities = S.fromList . concatMap getOutputsForRecipe $ rs ^. stdRecipes
nonCoveredEntities = Set.difference entityNames outputEntities
```